contracts/MarginCalculator.sol

```diff
@@ -2,11 +2,283 @@
+        OtokenInterface otoken = OtokenInterface(_otoken);
+        require(now > otoken.expiryTimestamp(), "MarginCalculator: Otoken not expired yet.");
+
+        uint256 strikePrice = otoken.strikePrice();
+        (uint256 underlyingPrice, bool isFinalized) = _getUnderlyingPrice(_otoken);
+        require(isFinalized, "MarginCalculator: Oracle price not finalized yet.");
+
+        if (otoken.isPut()) {
+            return strikePrice > underlyingPrice ? strikePrice.sub(underlyingPrice) : 0;
+        } else {
+            return underlyingPrice > strikePrice ? underlyingPrice.sub(strikePrice) : 0;
+        }
+    }
+
+    /**
+     * @notice returns the net value of a vault in the valid collateral asset for that vault i.e. USDC for puts/ ETH for calls
+     * @param _vault the theoretical vault that needs to be checked
+     * @param _denominated the token the result is denominated in. Must be the same as short.collateral for now.
+     * @return netValue the amount by which the margin is above or below the required amount.
+     * @return isExcess true if there's excess margin in the vault. In this case, collateral can be taken out from the vault. False if there is insufficient margin and additional collateral needs to be added to the vault to create the position.
+     */
+    function getExcessMargin(MarginAccount.Vault memory _vault, address _denominated)
+        public
+        view
+        returns (uint256 netValue, bool isExcess)
+    {
+        // ensure the number of collateral, long and short array is valid.
+        _checkAssetCount(_vault);
+        // ensure the long asset is valid for the short asset.
+        _checkLongAsset(_vault);
+
+        // collateral amount is always positive.
+        int256 collateralAmount = _vault.collateralAmounts.length > 0
+            ? FixedPointInt256.uintToInt(_vault.collateralAmounts[0])
+            : 0;
+
+        // Vault contains no short tokens: return collateral value.
+        if (_vault.shortOtokens.length == 0) return (collateralAmount.intToUint(), true);
+
+        // For the currenct version, ensure denominated == short.collateral
+        address shortCollateral = OtokenInterface(_vault.shortOtokens[0]).collateralAsset();
+        require(shortCollateral == _denominated, "MarginCalculator: Denomintated token should be short.collateral");
+
+        int256 netOtoken = _calculateOtokenNetValue(_vault);
+        // if netOtoken < 0, the long assets cannot cover the max loss of short assets in the vault.
+        // will need to check if collateral + netOtoken is greater than 0.
+        int256 totalValue = collateralAmount.add(netOtoken);
+        isExcess = totalValue >= 0;
+        netValue = totalValue.intToUint();
+    }
+
+    /**
+     * @notice Calculate the net value of long token + short token in a vault, denominated in collateral asset.
+     * @dev The vault passed in already pass amount array length = asset array length check.
```

Comment:
Can we have the formulas for calls vs Puts here? Otherwise its confusing to understand the math.

Reply:
But we have 4 different formulas, how can we include all of them well in the comment section?